apache/storm
# [STORM-2693] Storm heartbeats promotion

Storm now doesn't support large cluster[ for example thousand of supervisors] very well, for our production, topology submission/killing is very ineffective when cluster grows to be large, i checkout the heartbeats strategy and find that actually it can be promoted.

For heartbeats promotion:

1. Nimbus will not collect heartbeats info from zookeeper every scheduling round any more, instead, it will use directly a cache which is updated when a heartbeat reports
2. Heartbeats will be reported through supervisor RPC [supervisor will collect local heartbeats from local state store which is reported by node workers]
3. Separate the metrics data and heartbeats, which means that the new heartbeat will not contains metrics info, so it is very lightweight and efficient
4. Metrics data will still be reported to zookeeper, but only used for collecting UI stats[ UI stats is got from heartbeats cache with the old mode, we fetch it from zookeeper directly now ]

![heartbeats-promotion](https://user-images.githubusercontent.com/7644508/32034188-be12919c-b9d6-11e7-9740-631aef5ff4b4.png)

With this new mode, heartbeats will be reported very efficiently, for our production, there are about 30 workers per node/supervisor, so i mock the data and did a pressure test for nimbus heartbeats response time:
Test condition:
1. an 8core/16Gb memory virtual machine for nimbus
2. start numbers of threads for mocking the cluster node PRC requests
3. start a new client to collect the request/response time stats

![image](https://user-images.githubusercontent.com/7644508/32034309-88f3b210-b9d7-11e7-8798-139a7eca10de.png)

We can see that for a 1 second heartbeat report frequency, nimbus will support at least 2000 nodes, for our production, we set the worker heartbeats reporting interval to 5 seconds, so it means that we can have a 5* 2000 nodes for just one cluster

Because there is no need to collect all heartbeats data and compute alive/free slots for every scheduling round[ use a computed cache directly], we schedule topologies very efficiently[ only 2 minutes for 5000 topologies]

About robustness:
1. when nimbus collapse, workers works fine[ like the original ], when leader starts up, it will wait for a complete heartbeats for all nodes with a configured timeout and  start to work again, i also make the strategy pluggable, user can override the default one
2. when supervisor goes down, workers still workers fine,[ they will report heartbeats directly to nimbus through RPC], when supervisor goes up, it will just collect the heartbeats and reports to nimbus
3. when zk is unstable, it will not affect the heartbeats[which will cause workers all collapse for old mode] any more

This is my JIRA task: https://issues.apache.org/jira/browse/STORM-2693
This is the assignments promotion PR: https://github.com/apache/storm/pull/2319
